File: napkin/generic/basewindow.py
from PyQt5.QtCore import QCoreApplication, QSettings
from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QMainWindow, QDockWidget


class BaseWindow(QMainWindow):
    __WIN_GEO = 'WindowGeometry'
    __WIN_STATE = 'WindowState'

    def __init__(self):
        super(BaseWindow, self).__init__()
        self.setWindowTitle(QCoreApplication.applicationName())
        self.setDockNestingEnabled(True)
        self.__windowMenu = None
        # self.centralWidget().setVisible(False)

    def windowMenu(self):
        if not self.__windowMenu:
            self.__windowMenu = self.menuBar().addMenu('&Window')
        return self.__windowMenu

    def addDock(self, name, widget, area=Qt.TopDockWidgetArea):
        dock = QDockWidget()
        dock.setObjectName(name)
        dock.setWidget(widget)
        dock.setWindowTitle(name)
        action = self.windowMenu().addAction(name)
        action.setCheckable(True)
        action.setChecked(True)
        self.addDockWidget(area, dock)

    def showEvent(self, e):
        super(BaseWindow, self).showEvent(e)
        s = QSettings()
        v = s.value(self.__WIN_GEO)
        if v:
            self.restoreGeometry(v)
        v = s.value(self.__WIN_STATE)
        if v:
            self.restoreState(v)

    def closeEvent(self, e):
        super(BaseWindow, self).closeEvent(e)
        s = QSettings()
        s.setValue(self.__WIN_STATE, self.saveState())
        s.setValue(self.__WIN_GEO, self.saveGeometry())

File: napkin/src/generic/basewindow.cpp
#include "basewindow.h"
#include "autosettings.h"


napkin::BaseWindow::BaseWindow()
{
	setWindowTitle(QApplication::applicationName());
	setDockNestingEnabled(true);
	mWindowMenu = new QMenu("Window");
	menuBar()->addMenu(mWindowMenu);
}

QDockWidget* napkin::BaseWindow::addDock(const QString& name, QWidget* widget, Qt::DockWidgetArea area)
{
	QDockWidget* dock = new QDockWidget(this);
	dock->setObjectName(name);
	dock->setWidget(widget);
	dock->setWindowTitle(name);

	if (widget->objectName().isEmpty())
		widget->setObjectName(QString("%1_Widget").arg(name));

	QAction* action = mWindowMenu->addAction(name);
	action->setCheckable(true);
	action->setChecked(true);
	connect(action, &QAction::triggered, [dock, action]() { dock->setVisible(action->isChecked()); });
	addDockWidget(area, dock);
	return dock;
}

void napkin::BaseWindow::showEvent(QShowEvent* event)
{
	QWidget::showEvent(event);
	AutoSettings::get().restore(*this);
}

void napkin::BaseWindow::closeEvent(QCloseEvent* event)
{
	AutoSettings::get().store(*this);
	QWidget::closeEvent(event);
}

